api/server/middleware_auth.go

```diff
@@ -450,44 +367,82 @@ func (a *authMiddleware) parseParam(r *http.Request, param string) (string, erro
 	return "", fmt.Errorf("could not parse %s", param)
 }
 
-func (a *authMiddleware) parseSecret(
+// This functions makes it possible to secure the model of accessing the secret by allowing
+// the definition of secret access to come from the storage class, as done by CSI.
+func (a *authMiddleware) getSecretInformationInKubernetes(
 	specLabels, locatorLabels map[string]string,
-	fetchCOLabels bool,
 ) (*api.TokenSecretContext, error) {
-	if lsecrets.Instance().String() == lsecrets.TypeK8s && fetchCOLabels {
-		// For k8s fetch the actual annotations
-		pvcName, ok := locatorLabels[PVCNameLabelKey]
-		if !ok {
-			// best effort to fetch the secret
-			return parseSecretFromLabels(specLabels, locatorLabels)
-		}
-		pvcNamespace, ok := locatorLabels[PVCNamespaceLabelKey]
-		if !ok {
-			// best effort to fetch the secret
-			return parseSecretFromLabels(specLabels, locatorLabels)
-		}
+	// Get pvc location and name
+	// For k8s fetch the actual annotations
+	pvcName, ok := getVolumeLabel(PVCNameLabelKey, specLabels, locatorLabels)
+	if !ok {
+		return nil, fmt.Errorf("Unable to authenticate request due to not able to determine name of the pvc from the volume")
+	}
+	pvcNamespace, ok := getVolumeLabel(PVCNamespaceLabelKey, specLabels, locatorLabels)
+	if !ok {
+		return nil, fmt.Errorf("Unable to authenticate request due to not able to determine namespace of the pvc from the volume")
+	}
 
-		pvc, err := core.Instance().GetPersistentVolumeClaim(pvcName, pvcNamespace)
-		if err != nil {
-			return nil, err
-		}
-		secretName := pvc.ObjectMeta.Annotations[osecrets.SecretNameKey]
+	// Get pvc object
+	pvc, err := core.Instance().GetPersistentVolumeClaim(pvcName, pvcNamespace)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get PVC information from Kubernetes: %v", err)
+	}
 
-		if len(secretName) == 0 {
-			return parseSecretFromLabels(specLabels, locatorLabels)
-		}
-		secretNamespace := pvc.ObjectMeta.Annotations[osecrets.SecretNamespaceKey]
+	// Get storageclass for pvc object
+	sc, err := core.Instance().GetStorageClassForPVC(pvc)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get StorageClass information from Kubernetes: %v", err)
+	}
+
+	// Get secret namespace
+	secretNamespaceValue := sc.Parameters[osecrets.SecretNamespaceKey]
+	secretNameValue := sc.Parameters[osecrets.SecretNameKey]
+	if len(secretNameValue) == 0 && len(secretNamespaceValue) == 0 {
+		return &api.TokenSecretContext{}, nil
+	}
+
+	// Allow ${pvc.namespace} to be set in the storage class
+	namespaceParams := map[string]string{"pvc.namespace": pvc.GetNamespace()}
```

Comment:
The ResolveTemplate function is only using this as an input to the template parser. In other words, the result will be the same if secretNamespaceValue is "" since there is nothing to replace.